Clean up/refactor GraphNode and make it more flexible

Geometror opened the pull request:
Split from #67152. Depends on #79308.
This PR contains the rework of GraphNode.

Detailed changes:
- Split `GraphNode` into `GraphElement` and `GraphNode`
- Renamed `set/get_connection_input/output_x` methods to `set/get_port_input/output_x`
- Remove `get_port_input/output_height` as it was not necessary (basically returned the height of the slot of the given port which is the height of the child in that slot)
- GraphNode: `get_port_input/output_position` returns no longer the position premultiplied with the scale
- Revised custom port drawing (now achievable via overriding a function)
- Flexibility improvements to GraphNode's titlebar
	- It now consists of a HBoxContainer with a Label (accessible via `get_titlebar_hbox
	- StyleBox for body (slot area) and title bar (both with selected variants)
	- Use "GraphNodeTitleLabel" theme variation for title labels, remove the title offset theme constants
- Respective adjustments to the editor theme and default project theme
- Rename lots of variables to be more descriptive (e.g. icofs -> icon_offset, c -> child)
- Restructure the header and source files to follow a consistent scheme (i. a. order of member variable and method declarations, which were sometimes mixed)
- Fix custom port icon not being centered when it's bigger than the default icon
- Fix and improve documentation with regard to port/connection/slot terminology
- Fix comment style in various places
- [can be extracted if desired] VisualShader: Fix error label increasing the minimum size in an ugly way since it autowrap isn't enabled

Summersay415 commented on doc/classes/GraphEdit.xml:
```diff
@@ -48,8 +48,8 @@
-				    var port_size: Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
-				    var port_pos: Vector2 = in_node.get_position() + in_node.get_connection_output_position(in_port) - port_size / 2
+				    var port_size : Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
+				    var port_pos : Vector2 = in_node.get_position() + in_node.get_port_output_position(in_port) - port_size / 2
```
https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html#static-typing

YuriSizov replied:
Thanks for the review. This isn't really a part of this PR. It's from a commit this PR is based on. The issue was fixed in the original commit, this PR just needs to be rebased now.

Geometror commented:
Rebased.

AThousandShips commented:
You need to handle the compatibility breakages

YuriSizov commented:
I'll review it this week, but for now I wanted to mention that grid toggle icons shouldn't be a part of this. They were added (though originally missing) in https://github.com/godotengine/godot/pull/79308.

Geometror committed:
Clean up/refactor GraphNode and make it more flexible

Split GraphNode into GraphElement and GraphNode, add custom
titlebar, and adjust theming.

YuriSizov commented:
I fixed the icons and added some missing renames. I noticed some existing inconsistencies with closing/deleting, and unified them using the term closing, closable as it makes the most sense from the perspective of the graph itself. We can discuss this, but at least it's all consistent now and doesn't use both terms interchangeably.

From a quick look at visual shaders everything works fine. So I'll continue with my review.

PS. I think the names that we've picked for updated classes are good, but we can ask for more opinions.

YuriSizov commented on scene/resources/default_theme/default_theme.cpp:
```diff
@@ -699,40 +699,43 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
+	theme->set_stylebox("normal", "GraphNodeTitleLabel", memnew(StyleBoxEmpty));
+	theme->set_font("font", "GraphNodeTitleLabel", Ref<Font>());
+	theme->set_font_size("font_size", "GraphNodeTitleLabel", -1);
+	theme->set_color("font_color", "GraphNodeTitleLabel", Color(1, 1, 1));
+	theme->set_color("font_shadow_color", "GraphNodeTitleLabel", Color(0, 0, 0, 0));
+	theme->set_color("font_outline_color", "GraphNodeTitleLabel", Color(1, 1, 1));
```
These should probably use some reference colors defined above. Or define new colors if necessary.

YuriSizov commented on editor/editor_themes.cpp:
```diff
@@ -1911,45 +1917,43 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	theme->set_icon("port", "GraphNode", port_icon);
 
-	theme->set_font("title_font", "GraphNode", theme->get_font(SNAME("main_bold_msdf"), SNAME("EditorFonts")));
+	//TODO: @Geometror Change to theme variations.
+	theme->set_stylebox("state_machine_frame", "GraphNode", smgraphsb);
+	theme->set_stylebox("state_machine_selected_frame", "GraphNode", smgraphsbselected);
```
Do you want to do this now or in a follow-up?

Geometror replied:
I looked at the actual usage in the animation tree editor. They aren't used as a replacement/variation of the GraphNode's "body" stylebox, but for some kin of overlay. However, after observing the AnimationTree in our platformer demo I'm not quite sure whether this overlay is actually drawn at the moment. For now, I removed the comment.

YuriSizov commented on editor/editor_themes.cpp:
```diff
@@ -1911,45 +1917,43 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
+	theme->set_constant("port_h_offset", "GraphNode", 0);
 	theme->set_constant("separation", "GraphNode", 1 * EDSCALE);
 
-	theme->set_icon("close", "GraphNode", theme->get_icon(SNAME("GuiCloseCustomizable"), SNAME("EditorIcons")));
```
The icon itself should be removed now, nothing else uses it.

YuriSizov commented on editor/editor_themes.cpp:
```diff
@@ -1911,45 +1917,43 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
+	// GraphNode's title Label
+	theme->set_type_variation("GraphNodeTitleLabel", "Label");
+
+	theme->set_stylebox("normal", "GraphNodeTitleLabel", memnew(StyleBoxEmpty));
```
```suggestion
	theme->set_stylebox("normal", "GraphNodeTitleLabel", make_empty_stylebox(0, 0, 0, 0));
```

YuriSizov commented on doc/classes/GraphElement.xml:
```diff
@@ -0,0 +1,71 @@
+		GraphElement is a [Container] control that provides the basic functionality to be used inside a [GraphEdit] control (as a child), meaning it will be selectable and transformable by the graph editor.
+	</brief_description>
+	<description>
+		GraphElement allows to create your own custom nodes which can be used as a child of a [GraphEdit] node, but they won't be able to connect to other nodes.
+		To create a custom node that can connect to other nodes, extend [GraphNode].
```
```suggestion
		[GraphElement] allows to create custom elements for a [GraphEdit] graph. By default such elements can be selected, resized, and repositioned, but they cannot be connected. For a graph element that allows for connections see [GraphNode].
```

YuriSizov commented on doc/classes/GraphElement.xml:
```diff
@@ -0,0 +1,71 @@
+		</member>
+		<member name="position_offset" type="Vector2" setter="set_position_offset" getter="get_position_offset" default="Vector2(0, 0)">
+			The offset of the GraphElement, relative to the scroll offset of the [GraphEdit].
+			[b]Note:[/b] You cannot use the position offset directly, as [GraphEdit] is a [Container].
```
```suggestion
			[b]Note:[/b] You cannot use the position offset directly, as [GraphEdit] is a [Container].
```

This comes from the existing description, but this is wrong and confusing. GraphEdit is not a container, and I'm not entirely sure what is being said here. What kind of "use" is implied. So I'd suggest removing the line, unless you know what it means, then we can improve it.

Geometror replied:
I'm confused as well about this sentence. Let's remove it.

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the [Color] of the input connection [param port].
+				Returns the [Color] of the input connection [param port_idx].
```
```suggestion
				Returns the [Color] of the input port for the given [param port_idx].
```

Geometror replied:
I would use "with the given" instead of "for the given".

YuriSizov replied:
Sure!

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the position of the input connection [param port].
+				Returns the position of the input port [param port_idx].
```
```suggestion
				Returns the position of the input port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the corresponding slot index of the input connection [param port].
+				Returns the corresponding slot index of the input port [param port_idx].
```
```suggestion
				Returns the corresponding slot index of the input port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the type of the input connection [param port].
+				Returns the type of the input connection [param port_idx].
```
```suggestion
				Returns the type of the input port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the [Color] of the output connection [param port].
+				Returns the [Color] of the output connection [param port_idx].
```
```suggestion
				Returns the [Color] of the output port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the position of the output connection [param port].
+				Returns the position of the output port [param port_idx].
```
```suggestion
				Returns the position of the output port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the corresponding slot index of the output connection [param port].
+				Returns the corresponding slot index of the output port [param port_idx].
```
```suggestion
				Returns the corresponding slot index of the output port for the given [param port_idx].
```

YuriSizov commented on doc/classes/GraphNode.xml:
```diff
@@ -25,86 +34,72 @@
+			<param index="0" name="port_idx" type="int" />
 			<description>
-				Returns the type of the output connection [param port].
+				Returns the type of the output port [param port_idx].
```
```suggestion
				Returns the type of the output port for the given [param port_idx].
```

YuriSizov reviewed (changes requested):
Documentation and theme review.

Please when you make changes do not rebase against the master, otherwise it would be a hell to untangle what has been actually changed between force pushes. Alternatively, making it into separate commits for now is also fine.

YuriSizov commented on scene/gui/graph_node.h:
```diff
@@ -31,170 +31,113 @@
 
-	void set_draggable(bool p_draggable);
-	bool is_draggable();
+	int get_port_input_count();
+	Vector2 get_port_input_position(int p_port_idx);
+	int get_port_input_type(int p_port_idx);
+	Color get_port_input_color(int p_port_idx);
+	int get_port_input_slot(int p_port_idx);
```
I think it would be better to call these `input/output_port` rather than `port_input/output`. For example, the name `get_port_input_count` doesn't read the way it should for what it returns. It reads like it returns some input count.

With the old names using "connection" it kind of worked, because it was "connection input". Where as with "port" the word "input" it its qualifier. In other words, connection input is input port.

Current format also adds to the whole slot/port confusion, I think. When you read "port input color", it feels like what "slot input color" would mean. An input color of something, implying there is also an output color of the same thing. Whereas "input port color" is more precise and less misleading.

YuriSizov commented on scene/gui/graph_node.h:
```diff
@@ -31,170 +31,113 @@
 
-	void set_selectable(bool p_selectable);
-	bool is_selectable();
+	int get_port_output_count();
+	Vector2 get_port_output_position(int p_port_idx);
+	int get_port_output_type(int p_port_idx);
+	Color get_port_output_color(int p_port_idx);
+	int get_port_output_slot(int p_port_idx);
```
Same applies to output, of course.

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+}
+#endif
+
+void GraphElement::_close_requested() {
+	// Send focus to parent.
+	get_parent_control()->grab_focus();
+	emit_signal(SNAME("close_request"));
+}
```
This method is not virtual and is not called anywhere, why do we need it? We can (and probably should) keep the signal, but the method seems redundant.

Geometror replied:
Seems like this method wasn't even used before. Will remove it.

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+		Size2i size = child->get_combined_minimum_size();
+
+		minsize.x = MAX(minsize.x, size.x);
+		minsize.y += MAX(minsize.y, size.y);
```
```suggestion
		minsize.y = MAX(minsize.y, size.y);
```

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+}
+
+void GraphElement::set_position_offset(const Vector2 &p_offset) {
+	position_offset = p_offset;
```
This used to have a check and return early without any assignment if the new value is the same. I think we should keep it, since the method has side-effects.

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+}
+
+void GraphElement::set_resizable(bool p_enable) {
+	resizable = p_enable;
```
Same here.

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+	return draggable;
+}
+
+void GraphElement::set_selectable(bool p_selectable) {
+	if (!p_selectable) {
+		set_selected(false);
+	}
+	selectable = p_selectable;
+}
```
A similar check can probably also be added here.

Geometror replied:
There is already a check in set_selected and this method has no additional side effects, so I wouldn't add another check here.

YuriSizov commented on scene/gui/graph_element.cpp:
```diff
@@ -0,0 +1,244 @@
+	return drag_from;
+}
+
+void GraphElement::gui_input(const Ref<InputEvent> &p_ev) {
```
This method is still present in `GraphNode` verbatim. Should it be removed from there?

Geometror replied:
I think so, it can be added back in case we need to add specific behavior. Currently, the only difference is the error message (which will be a bit less precise when the GraphNode implementation of the method is removed, but that's ok IMO).

YuriSizov commented on scene/gui/graph_node.cpp:
```diff
@@ -139,87 +150,91 @@ void GraphNode::_get_property_list(List<PropertyInfo> *p_list) const {
-				continue;
-			}
-			if (c->is_set_as_top_level()) {
+	// Second, pass successively to discard elements that can't be stretched, this will run
+	// while stretchable elements exist.
```
```suggestion
	// Second pass, discard elements that can't be stretched, this will run while stretchable elements exist.
```

YuriSizov replied:
You changed it to be a single line, but that's not what this suggestion was about. 🙂  I proposed different phrasing, which also fixes some weird grammar/punctuation.

Geometror replied:
Whoops, fixed.

YuriSizov commented on scene/gui/graph_node.cpp:
```diff
@@ -228,697 +243,493 @@ void GraphNode::_resort() {
-		} else {
-			ofs += sep;
+		// Adjust so the last one always fits perfect, compensating for numerical imprecision.
+		if (msc.will_stretch && i == children_count - 1) {
```
This doesn't work since `children_count` only counts valid children, and `i` enumerates all children. The old `idx` variable seems to be useful after all.

YuriSizov commented on scene/gui/graph_node.cpp:
```diff
@@ -228,697 +243,493 @@ void GraphNode::_resort() {
+
+		_MinSizeCache &msc = min_size_cache[child];
+
+		if (i > 0) {
+			ofs_y += separation;
```
And same here.

YuriSizov reviewed (changes requested):
Code review. As with `GraphEdit` it's still missing the manual theme cache, but I guess I'll add it after all this rework.

We should also add compatibility exceptions for the new renames and other public API changes, see https://github.com/godotengine/godot/pull/80168. I plan to follow this up with a PR that adds compatibility methods where possible, but for now adding exceptions should be fine.

YuriSizov commented:
I haven't checked the code too closely, but it seems mostly fine outside of outlined problems. This is to say that once those are addressed, I think we can go ahead with this, and use public testing to help us identify if there remain any typos or calculation mistakes.

What remains to test is using GraphEdit in a project, customizing it, extending elements and configuring the titlebar. I will do this tomorrow.

Geometror commented on scene/gui/graph_node.h:
```diff
@@ -31,170 +31,113 @@
+class GraphNode : public GraphElement {
+	GDCLASS(GraphNode, GraphElement);
 
 private:
```
I thought this was somewhat code-style related. We do this is many classes (e.g. Label, Button) so I suggest keeping it for now. If we decide that the redundant `private` modifier should be removed, then I would do it for in a cleanup PR for all classes.

YuriSizov replied:
I haven't checked those classes, but we don't normally add private if it's the first block. Here it's there because there used to be a block with a different scope above it.

YuriSizov commented:
We should also remove the configuration warning with this PR:

![image](https://github.com/godotengine/godot/assets/11782833/adcaf872-0181-4584-9a1c-29cac1dfe72f)

YuriSizov commented:
Tested functionality. The improvements are very nice! I am a bit concerned that it's impossible to make something closable without writing code that adds buttons, but I guess you are free to define your own subclass of `GraphNode` that does that. We can also see if there is any negative feedback to this change.

There are a couple of sizing issues.

1. One I've noticed during the code review already, where child's height was added during the minimum size calculation. This leads to `GraphElement`s growing in height the more children they have.

2. For the first frame `GraphNode`'s titlebar has incorrect dimensions. It basically fits only the label itself and doesn't account for stylebox margins or other children. It fixes itself if you click on it.

<img width="579" alt="NVIDIA_Share_2023-08-10_14-26-46" src="https://github.com/godotengine/godot/assets/11782833/cde9e580-9c22-4df0-8bd7-34f555952216">
<img width="578" alt="NVIDIA_Share_2023-08-10_14-27-19" src="https://github.com/godotengine/godot/assets/11782833/4839a887-7dc3-4a4a-9540-967d6eb5cb67">

3. Resizable `GraphNode`s can't shrink under certain conditions. If you resize it and let go, then resize it again, it all works fine. However, if you click anywhere (on the node itself, on the title, on the empty space) and then try to resize again, the current size is going to stay fixed.

https://github.com/godotengine/godot/assets/11782833/98d83753-4c44-4b12-bab9-274e18b2aa8f

Geometror commented:
Committed the fixes/changes, should address all review comments.
I would remove the configuration warning in a few weeks after making sure that everything works fine and we haven't overlooked any major design flaw.
The highlights for the menu buttons in the default theme aren't introduced with this or recent PRs (already present in 4.0.4), so I would rather not include it in this one, but in a follow-up.

YuriSizov reviewed (approved):
LGTM overall. Great job and sorry for the delay with confirming changes! There is still one suggestion that you haven't fully addressed, but it should be a quick fix.

I also noticed another issue, but it should also be a quick fix. Snapping to the grid only works with position, but not size. Even with snapping enabled, you can change the size of graph nodes freely. If this proves to be not-so-trivial to fix, I think we should be fine with accepting this regression, merging, and fixing the bug as a follow-up.

You still need to add exceptions to the API exceptions file for the renamed/removed properties and methods. I've linked to another PR before that should demonstrate how it can be done. We can fill the compatibility methods in a follow-up (in fact, that would be preferrable).

Once you address those issues, feel free to squash and rebase your PR for other reviewers!

YuriSizov commented:
> More generally, I also have lots of requests about UI interactions with connections (being able to insert a node into a connection, select a connection to remove it etc.), mainly inspired by Blender. Not sure which ones actually require changes in Godot, though.

I think GraphEdit would need to provide a utility method/signal to detect clicks on the connection line. And the rest can be done on the application side.

Chaosus commented:
Looks like visual shader nodes now misses the delete button - does it intended?
![image](https://github.com/godotengine/godot/assets/3036176/1396a7aa-242d-4fcc-a98d-e1782c4f3df5)

Geometror commented:
@RodZill4
> I'd really like the "new comment node" to be able to contain any Control (something like a Window pinned to the graph canvas). Would be much more flexible and not limited to comments.

This will definitely be the case. You can probably implement a basic comment node right now (without fancy sorting or automatic resizing) using the new base class `GraphElement`.

![ge3](https://github.com/godotengine/godot/assets/50084500/a274c6cb-2f3e-4608-8a91-ae3d6f5ca7ee)

> More generally, I also have lots of requests about UI interactions with connections (being able to insert a node into a connection, select a connection to remove it etc.), mainly inspired by Blender. Not sure which ones actually require changes in Godot, though.

I have plans to construct an API around connections, but this has to wait for 4.3+.

YuriSizov commented:
> This is a bit more verbose, but doing input checks in `graph_element_resized` feels not right, so I went with this approach.

I think singleton exists for this specific purpose, so you can check the state of the input at any point. So I don't have a problem with just copying the logic from `GraphEdit::gui_input`. But I'm not against setting a flag either. Just as long as we don't end up with many of them, making it's harder to understand the current state of the node (because many flags mean many permutations of them).

AThousandShips commented on doc/classes/GraphEdit.xml:
```diff
@@ -289,6 +289,12 @@
+		<signal name="close_nodes_request">
+			<param index="0" name="nodes" type="StringName[]" />
+			<description>
+				Emitted when a GraphNode is attempted to be removed from the GraphEdit. Provides a list of node names to be removed (all selected nodes, excluding nodes without closing button).
```
```suggestion
				Emitted when attempting to remove a GraphNode from the GraphEdit. Provides a list of node names to be removed (all selected nodes, excluding nodes without closing button).
```

AThousandShips commented on doc/classes/GraphElement.xml:
```diff
@@ -0,0 +1,69 @@
+	<signals>
+		<signal name="close_request">
+			<description>
+				Emitted when the GraphElement is requested to be closed.
```
```suggestion
				Emitted when closing the GraphElement is requested.
```

AThousandShips commented on doc/classes/GraphElement.xml:
```diff
@@ -0,0 +1,69 @@
+		</signal>
+		<signal name="raise_request">
+			<description>
+				Emitted when the GraphElement is requested to be displayed over other ones. Happens on focusing (clicking into) the GraphElement.
```
```suggestion
				Emitted when displaying the GraphElement over other ones is requested. Happens on focusing (clicking into) the GraphElement.
```

AThousandShips commented on doc/classes/GraphElement.xml:
```diff
@@ -0,0 +1,69 @@
+		<signal name="resize_request">
+			<param index="0" name="new_minsize" type="Vector2" />
+			<description>
+				Emitted when the GraphElement is requested to be resized. Happens on dragging the resizer handle (see [member resizable]).
```
```suggestion
				Emitted when resizing the GraphElement is requested. Happens on dragging the resizer handle (see [member resizable]).
```

AThousandShips commented on misc/extension_api_validation/4.1-stable.expected:
```diff
@@ -116,3 +116,54 @@ Validate extension JSON: API was removed: classes/GLTFDocumentExtensionPhysics
+Validate extension JSON: API was removed: classes/GraphNode/signals/node_selected
+Validate extension JSON: API was removed: classes/GraphNode/signals/position_offset_changed
+Validate extension JSON: API was removed: classes/GraphNode/signals/raise_request
+Validate extension JSON: API was removed: classes/GraphNode/signals/resize_request
```
```suggestion
Validate extension JSON: API was removed: classes/GraphNode/signals/resize_request

```
Missing line termination

AThousandShips reviewed (approved):
Other than the grammar fixes above this LGTM

AThousandShips commented:
No worries! Next time try out the "Add suggestion to batch" option to do all in one commit

akien-mga commented:
Thanks!

YuriSizov commented:
> @YuriSizov The snapping was already fixed (when resizing).

I don't see any relevant changes and you haven't pushed anything before I asked the last time 👀

YuriSizov commented:
@Geometror Just to re-confirm, because I've just tested it on master, but there is no snapping when resizing. I think you haven't pushed the fixed that you've described before.

nonchip commented:
> We should also remove the configuration warning with this PR:

@YuriSizov btw that never happened, might be worth doing before the upcoming release?

AThousandShips commented:
`GraphEdit` is still experimental so I'd keep it until that changes

AThousandShips commented:
Leaving it as experimental gives us room to tweak and fix, my suggestion is to leave it until 4.2.1 at the earliest to get some more field testing of it done

YuriSizov commented:
> > We should also remove the configuration warning with this PR:
> 
> @YuriSizov btw that never happened, might be worth doing before the upcoming release?

We decided to keep it marked as experimental until 4.3, but there shouldn't be major breaking changes. It's just to account for possible feedback from GraphEdit users, who aren't many.

nonchip commented:
@YuriSizov ah yeah that *kinda* makes sense, the main issue i see with it is just that currently the most common feedback is "i'm scared to become a GraphEdit user, because it literally tells me to wait for an undetermined amount of time before i do so, when will the rework finally happen so i can start using it?"

RodZill4 commented:
Updating Material Maker for Godot 4.2 has been pretty smooth on the GraphEdit/GraphNode side. Didn't notice any problem (and graph nodes are used A LOT). The only aspect I didn't dig into yet is converting comment nodes into GraphElements.

Geometror commented:
@RodZill4 Comment nodes are likely going to be difficult to implement just with `GraphElement` (at least how they used to work) due to the removal of dedicated logic in `GraphEdit`. Comment nodes were removed because the implementation was problematic and I couldn't come up with a better approach in time. I had a few prototypes which did *almost* work (but failed in certain cases). This has a high priority for me, but time is a scarce commodity and I'd like to merge the other `GraphEdit` related PRs first, since they might alter the implementation of `GraphFrame` (the dedicated comment node class).

nagidev commented:
@Geometror can you please take a look at [this issue](https://github.com/godotengine/godot/issues/85558) which was probably introduced due to this pull? I'm not familiar with the codebase, so I could do with some help understanding the `GraphNode::_resort` function.
